spec.html

```diff
@@ -7875,6 +7880,251 @@ <h1>Forward Progress</h1>
+        not imply that the object is live. For example if the object in
+        question is never passed back to the program, then it cannot be
+        observed.
+        </p>
+
+        <p>
+        This is the case for keys in a WeakMap, members of a WeakSet, as well
+        as the [[WeakRefTarget]] and [[UnregisterToken]] fields of a FinalizationRegistry
+        Cell record.
+        </p>
+
+        <p>
+        The above definition implies that, if a key in a WeakMap is not live,
+        then its corresponding value is not necessarily live either.
+        </p>
+      </emu-note>
+      <emu-note>
+        Liveness is the lower bound for guaranteeing which WeakRefs that
+        engines must not empty. In practice, liveness as defined here is
+        undecidable and engines use conservative approximations. There is
+        expected to be significant implementation leeway.
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-weakref-execution">
+      <h1>Execution</h1>
+
+      <p>At any time, if a set of objects _S_ is not live, an ECMAScript implementation may perform the following steps atomically:</p>
+
+      <emu-alg>
+        1. For each _obj_ of _S_, do
+          1. For each WeakRef _ref_ such that _ref_.[[WeakRefTarget]] is _obj_, do
+            1. Set _ref_.[[WeakRefTarget]] to ~empty~.
+          1. For each FinalizationRegistry _fg_ such that _fg_.[[Cells]] contains _cell_, such that _cell_.[[WeakRefTarget]] is _obj_, do
+            1. Set _cell_.[[WeakRefTarget]] to ~empty~.
+            1. Optionally, perform ! HostCleanupFinalizationRegistry(_fg_).
+          1. For each WeakMap _map_ such that _map_.[[WeakMapData]] contains a record _r_ such that _r_.[[Key]] is _obj_, do
+            1. Remove _r_ from _map_.[[WeakMapData]].
+          1. For each WeakSet _set_ such that _set_.[[WeakSetData]] contains _obj_, do
+            1. Remove _obj_ from _set_.[[WeakSetData]].
+      </emu-alg>
+
+      <emu-note>
+        <p>
+          Together with the definition of liveness, this clause prescribes legal
+          optimizations that an implementation may apply regarding WeakRefs.
+        </p>
+
+        <p>
+          It is possible to access an object without observing its
+          identity. Optimizations such as dead variable elimination, and scalar
+          replacement on properties of non-escaping objects whose identity is not
+          observed, are allowed. These optimizations are thus allowed to
+          observably empty WeakRefs that point to such objects.
+        </p>
+
+        <p>
+          On the other hand, if an object's identity is observable, and that
+          object is in the [[WeakRefTarget]] internal slot of a WeakRef, optimizations
+          such as rematerialization that observably empty the WeakRef are
+          prohibited.
+        </p>
+
+        <p>
+          Because calling HostCleanupFinalizationRegistry is optional, registered objects
```

Comment:
```suggestion
          Because calling CleanupFinalizationRegistry is optional, registered objects
```

Reply:
I also disagree with this suggestion for the same reason as above.